### DISCUSSION TITLE
Automatic reconnection when peripheral disconnects

### DISCUSSION THREAD

**Discussion by ghost** [hbldh/bleak#478] [Q&A]
* bleak version: 0.10.0
* Python version: 3.7.3
* Operating System: Windows 10

### Description

On Windows, a bleak client will automatically reconnect when the peripheral disconnects. This is undesirable behaviour in cases where it is required to completely terminate the connection with the host. For example, consider a `DisconnectionService` with a      
a single characteristic: `disconnection_reason_char`. If a valid disconnection reason is written to the characteristic, the service immediately disconnects from the host. Such a service is useful when writing integration tests for BLE devices.     

### What I Did

I wrote a BLE application using [mbed-os](https://os.mbed.com/mbed-os/) that contains a `DisconnectionService`. Then, I built and flashed the application onto a [NUCLEO-WB55RG](https://os.mbed.com/platforms/ST-Nucleo-WB55RG/); however, any Mbed BLE-enabled [development board](https://os.mbed.com/platforms/?q=&Communication=BLE) should work. Lastly, I wrote a Python script to interact with the service using bleak.

The main `run` function is shown below. It scans for and attempts to connect to a device named "Disconnection", viz. the name of my peripheral. Once connected, it writes a valid disconnection reason to `disconnection_reason_char` of the `DisconnectionService`. Therefore, the peripheral immediately disconnects from the host.       

```python
async def run():
    devices = await BleakScanner.discover()
    for d in devices:
        if d.name == "Disconnection":
            client = BleakClient(d)
            connected = await client.connect()
            if connected:
                await client.write_gatt_char(UUID_DISCONNECTION_REASON_CHAR, USER_TERMINATION)
                time.sleep(60)
                connected = await client.is_connected()
                l.info(f'Connected: {connected}')
```

I set the log level to DEBUG and observed traces from bleak (see below). We see that `_ConnectionStatusChanged_Handler` gets flipped to 0 after `disconnection_reason_char` is written with `0x13`. This is expected because that corresponds to user termination, a valid disconnection reason. However, ~15 seconds later `_ConnectionStatusChanged_Handler` gets flipped to 1; and, after the 1 minute pause, we see that the client has reconnected with the peripheral.       

```
2021-03-08 15:43:52,832.832 DEBUG Using selector: SelectSelector
2021-03-08 15:43:53,442.442 DEBUG Received E9:1E:1F:57:F9:4C: Disconnection.
2021-03-08 15:43:53,443.443 DEBUG Received xx:xx:xx:xx:xx:xx: Unknown.
2021-03-08 15:43:55,516.516 DEBUG Received xx:xx:xx:xx:xx:xx: Unknown.
2021-03-08 15:43:55,517.517 DEBUG Received xx:xx:xx:xx:xx:xx: Unknown.
2021-03-08 15:43:56,876.876 DEBUG Received xx:xx:xx:xx:xx:xx: Unknown.
2021-03-08 15:43:56,877.877 DEBUG Received xx:xx:xx:xx:xx:xx: Unknown.
2021-03-08 15:43:57,899.899 DEBUG Connecting to BLE device @ E9:1E:1F:57:F9:4C
2021-03-08 15:44:04,561.561 DEBUG _ConnectionStatusChanged_Handler: 1
2021-03-08 15:44:04,563.563 DEBUG Get Services...
2021-03-08 15:44:05,954.954 INFO Services resolved for BleakClientDotNet (E9:1E:1F:57:F9:4C)
2021-03-08 15:44:06,077.077 DEBUG Write Characteristic f43620d0-779d-11eb-9439-0242ac130002 : bytearray(b'\x13')
2021-03-08 15:44:06,253.253 DEBUG _ConnectionStatusChanged_Handler: 0
2021-03-08 15:44:16,691.691 DEBUG _ConnectionStatusChanged_Handler: 1
2021-03-08 15:45:06,077.077 INFO Connected: True
```  

#### Compiling and running the Mbed BLE application
1. Clone and enter the repository:

    ```
    git clone https://github.com/noonfom/mbed-os-example-disconnection-service
    cd mbed-os-example-disconnection-service
    ```

1. Link dependencies:

    ```
    mbed deploy
    ```

1. Build and flash for a given `<target>` and `<toolchain>`:

    ```
    mbed compile -t <target> -m <toolchain> -f
    ```

### Attachments
[automatic_reconnection_log.txt](https://github.com/hbldh/bleak/files/6103014/automatic_reconnection_log.txt)

P.S. apologies for all the Mbed-specific stuff.

**Comment by dlech** [comment]
> ```python
>                 time.sleep(60)
> ```

This is blocking the async event loop from running. Use `await asyncio.sleep(60)` instead.

**Reply by ghost** [reply]
@dlech Thanks for the quick reply. I applied your suggestion and it fixed the issue :tada:
However, let's say the blocking function is not `time.sleep()` but a `wait_for_output()` function that waits for a fixed period of time until a specific response is received on the serial line. What would you advise then?

**Reply by dlech** [reply]
There is https://pyserial-asyncio.readthedocs.io, but I don't think it works on Windows. 

Another option is to use something like [loop.run_in_executor()](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor) to run the blocking task in a background thread.

**Reply by dlech** [reply]
Or, based on your description, `asyncio.Event` might be more appropriate to get a signal from a background thread.
